### PR TITLE
Issue 1908 - Fwd definition of beta function not always called

### PR DESCRIPTION
**Pull request by andrjohns**
## Summary

This PR fixes the issue where distributions that call the beta function with a namespace qualifier (i.e., ```stan::math::beta(a,b)```) don't delegate to the ```fwd``` definition for ```fvar``` inputs, rather relying on autodiff-ing the ```prim``` function.

By removing the namespace qualifier from the call to beta, the correct (```fwd```) definition is called.

## Tests
No tests added, as this PR fixes existing failures.

As shown in #1908, current tests will fail when the ```prim``` definition of ```beta``` is restricted to only take arithmetic inputs. This PR introduces that restriction, and tests pass with the added changes.

## Side Effects

N/A

## Release notes

Fixed a bug where the appropriate specialization for the beta function was not called for forward-mode autodiff variables

## Checklist

- [x] Math issue #1908

- [x] Copyright holder: Andrew Johnson

    The copyright holder is typically you or your assignee, such as a university or company. By submitting this pull request, the copyright holder is agreeing to the license the submitted work under the following licenses:
      - Code: BSD 3-clause (https://opensource.org/licenses/BSD-3-Clause)
      - Documentation: CC-BY 4.0 (https://creativecommons.org/licenses/by/4.0/)

- [x] the basic tests are passing

    - unit tests pass (to run, use: `./runTests.py test/unit`)
    - header checks pass, (`make test-headers`)
    - dependencies checks pass, (`make test-math-dependencies`)
    - docs build, (`make doxygen`)
    - code passes the built in [C++ standards](https://github.com/stan-dev/stan/wiki/Code-Quality) checks (`make cpplint`)

- [x] the code is written in idiomatic C++ and changes are documented in the doxygen

- [x] the new changes are tested

### CODE REVIEW

File: stan/math/prim/fun/beta.hpp [hunk 1]
```diff
@@ -48,7 +48,7 @@ namespace math {
  * @param b Second value
  * @return Beta function applied to the two values.
  */
-template <typename T1, typename T2>
+template <typename T1, typename T2, require_all_arithmetic_t<T1, T2>* = nullptr>
```
Comment (bob-carpenter) [inline stan/math/prim/fun/beta.hpp:51]: [question]
I'm confused as to why we have seemingly three different ways to do this:

1.  `typename = require_all_arithmetic_t<...>`
2. `require_all_arithmetic_t<...>...`
3. `require_all_arithmetic_t<T1, T2>* = nullptr`

I'd rather just settle on one unless there's an important difference.  @SteveBronder?

File: stan/math/prim/prob/beta_lccdf.hpp [hunk 2]
```diff
@@ -29,43 +29,43 @@ namespace math {
...
+ * @param beta_param (Sequence of) failure parameter(s)
  * @return log probability or sum of log of probabilities
  * @throw std::domain_error if alpha or beta is negative
  * @throw std::domain_error if y is not a valid probability
  * @throw std::invalid_argument if container sizes mismatch
  */
 template <typename T_y, typename T_scale_succ, typename T_scale_fail>
 return_type_t<T_y, T_scale_succ, T_scale_fail> beta_lccdf(
-    const T_y& y, const T_scale_succ& alpha, const T_scale_fail& beta) {
+    const T_y& y, const T_scale_succ& alpha, const T_scale_fail& beta_param) {
```
Comment (bob-carpenter) [inline stan/math/prim/prob/beta_lccdf.hpp:40]: `alpha` and `beta` are the same type of argument here, so either they should both get decorated with `_param` or neither should.  I'd prefer not to decorate conventional names like this, but I won't complain if you decide to add `_param` to `alpha`, too.  Shouldn't `y` then also be decorated with `y_outcome` or something like that?

File: stan/math/prim/prob/neg_binomial_lccdf.hpp [hunk 3]
```diff
@@ -93,7 +93,7 @@ return_type_t<T_shape, T_inv_scale> neg_binomial_lccdf(
     const T_partials_return p_dbl = beta_dbl * inv_beta_p1;
     const T_partials_return d_dbl = square(inv_beta_p1);
     const T_partials_return Pi = 1.0 - inc_beta(alpha_dbl, n_dbl + 1.0, p_dbl);
-    const T_partials_return beta_func = stan::math::beta(n_dbl + 1, alpha_dbl);
+    const T_partials_return beta_func = beta(n_dbl + 1, alpha_dbl);
```
Comment (bob-carpenter) [inline stan/math/prim/prob/neg_binomial_lccdf.hpp:96]: [question]
Any idea why the right version wasn't getting picked up with the qualifier?  It should have all the candidates in scope as long as they've been included and if they haven't been included, I don't see how the unqualified version could work.

**Comment by bob-carpenter** [comment]
Nothing stopping the merge here, so I approved.  Feel free to merge when tests pass.

File: stan/math/prim/fun/beta.hpp (hunk 1 above)
Comment (SteveBronder) [reply stan/math/prim/fun/beta.hpp:51]: Let's stick with `require_all_arithmetic_t<T1, T2>* = nullptr`

Comment (bob-carpenter) [reply stan/math/prim/fun/beta.hpp:51]: I'm OK with any of (1) -- (3).  Should we always use the `ptr = nullptr` idiom?  I was wondering if there were differences in the approaches.

Comment (t4c1) [reply stan/math/prim/fun/beta.hpp:51]: 1. `typename = ` cannot be used for function overloads that opnly differ in requires. 
2. `...` can not be used on classes that need more than one require. Although I don't think we use requires for classes much.
3. `* = nullptr` has AFAIK no such restrictions. 

So either 2. or 3. is fine. 2. is the shortest, but so far we have used 3. the most.

**Pull request merged**